cluster-autoscaler/cloudprovider/azure/azure_scale_set.go

```diff
@@ -558,11 +561,18 @@ func (scaleSet *ScaleSet) waitForDeleteInstances(future *azure.Future, requiredI
 	isSuccess, err := isSuccessHTTPResponse(httpResponse, err)
 	if isSuccess {
 		klog.V(3).Infof(".WaitForDeleteInstancesResult(%v) for %s success", requiredIds.InstanceIds, scaleSet.Name)
-		// No need to invalidateInstanceCache because instanceStates were proactively set to "deleting"
+		if scaleSet.manager.config.StrictCacheUpdates {
+			if err := scaleSet.manager.forceRefresh(); err != nil {
+				klog.Errorf("forceRefresh failed with error: %v", err)
+			}
+			scaleSet.invalidateInstanceCache()
+		}
 		return
 	}
-	// On failure, invalidate the instanceCache - cannot have instances in deletingState
-	scaleSet.invalidateInstanceCache()
+	if !scaleSet.manager.config.StrictCacheUpdates {
+		// On failure, invalidate the instanceCache - cannot have instances in deletingState
+		scaleSet.invalidateInstanceCache()
```

Comment:
Should we add cache refresh here as well?
```suggestion
		if err := scaleSet.manager.forceRefresh(); err != nil {
				klog.Errorf("forceRefresh failed with error: %v", err)
		}
		scaleSet.invalidateInstanceCache()
```

Reply:
Doing so would introduce a functional change for folks not opting into this new config feature. I think keeping the current behaviors unmodified for the default use case is a priority.